### PR TITLE
Replace year with a full date in movies and musicvideos

### PR DESCRIPTION
**Pull request by phate89**
This allows to store the full date of a movie and a musicvideo.
I left the "year" label because it's the same for tv show. 
Since there's no way to rescan all the dates with the scrapers (or at least I think so) I set the current dates to the 1st January of the existing year.

edit:
actualy thinking about it I'm not sure if it's the right way to go. 
There are 2 ways:

this implementation. One field in the database and extract the year from it. This should be a no brainer but the problem is that we don't have the full date and the only way to update current movies is to rescrape the library. skins have no way to distinguish if the date is good or is a dummy year-01-01
Add a second field with the full date. In this way skins could show the full date only if it's a true full date. The problem is that we will have to keep 2 fields with the same info and keep 2 different variables in check in CVideoInfoTag for a long time.
Change between this 2 version is easy.. What do you think it's best?

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
How will this be shown in the gui after this change?

**Comment by HitcherUK** [comment]
Do the scrapers pull the full release date or just the year?

Also, from a skinning prospective, would be nice if we could format the label in the same way `System.Date(format)` works.

```Show current date using format, available markings: d (day of month 1-31), dd (day of month 01-31), ddd (short day of the week Mon-Sun), DDD (long day of the week Monday-Sunday), m (month 1-12), mm (month 01-12), mmm (short month name Jan-Dec), MMM (long month name January-December), yy (2-digit year), yyyy (4-digit year).```

**Comment by phate89** [comment]
@MartijnKaijser 
The same as before only with an option to show the full date.
Right now there is VIDEOPLAYER/LISTITEM_YEAR and VIDEOPLAYER/LISTITEM_PREMIERED but VIDEOPLAYER/LISTITEM_PREMIERED are empty for movies and musicvideos. Now they have full date and the year label behave as before
@HitcherUK 
right now we have only year in the database and no scraper support so all the dates in PREMIERED for movies and musicvideos are existing_year-01-01. With updated scrapers we will be able to get the full date (if the information provider supports it) but current release dates will stay the same..
Right now dates are formatted like for tv shows.. with localized date..

**Comment by Razzeee** [comment]
Sounds, good haven't read any code on it yet.
But Ping @olympia so he knows of this plan.

**Comment by HitcherUK** [comment]
Can we keep `foo.Year` for just displaying the year then?

**Comment by phate89** [comment]
actualy thinking about it I'm not sure if it's the right way to go. 
There are 2 ways:

1. this implementation. One field in the database and extract the year from it. This should be a no brainer but the problem is that skins have no way to distinguish if the date is good or is a dummy year-01-01
2. Add a second field with the full date. In this way skins could show the full date only if it's a true full date. The problem is that we will have to keep 2 fields with the same info for a long time.

Change between this 2 version is easy.. What do you think it's best?

@HitcherUK I didn't drop year label so unless I did something wrong the year label is still there only pulled from the full date

File: xbmc/video/VideoDatabase.cpp [hunk 1]
```diff
@@ -8698,8 +8695,8 @@ void CVideoDatabase::ExportToXML(const std::string &path, bool singleFile /* = t
         if (singleFile)
         {
           std::string strFileName(StringUtils::Join(movie.m_artist, g_advancedSettings.m_videoItemSeparator) + "." + movie.m_strTitle);
-          if (movie.m_iYear > 0)
-            strFileName += StringUtils::Format("_%i", movie.m_iYear);
+          if (movie.m_premiered.IsValid())
+            strFileName += StringUtils::Format("_%i", movie.m_premiered.IsValid());
```
Comment (anaconda) [inline xbmc/video/VideoDatabase.cpp]: `info.m_premiered.GetYear()`

**Comment by Montellese** [comment]
Looks good AFAICT. JSON-RPC changes should be safe.

File: xbmc/video/VideoDatabase.cpp (hunk 1 above)
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: Good catch..I'll fix it

**Comment by Tolriq** [comment]
Just 2 side notes :
- If both field stays linked it should be added somewhere in the API definition so that API consumers understand why changing one value update the other (specially if they try to update both at the same time)
- Year and premiered should be differentiated, Year is often seen as the official movie release date in the user country while premiered is the real premiered of the movie. It could pass some times between both date and so difference in Year value.

**Comment by phate89** [comment]
@Tolriq  about the first case sure.. But I don't know where I can state that..
The 2nd thing why exactly? First of all year is the part that already exists. It's the same as before.. Nothing changed. Contains the same information as before. Second thing both can be misunderstood as the local premiere date/year. Not only year. Plus it's consistent with the rest of the api. Tvshow is already handled in the same way. Year is extracted from the premiered date..

**Comment by Tolriq** [comment]
Well because I'm French :) And until recently most of the time movies get's here later than on original country. Same problem with movies that are presented at Festival de Cannes but really released later.

So IMO it makes sense to have a Year that can be unrelated to premiered. Premiered have the definition in it's name, there's only one premiere, else it's second ;)

But this was just a remark about a user case.

**Comment by phate89** [comment]
I'm italian so I see same strange dates but that's not true. There's the italian premiere, there's the french premiere, us premiere etc etc Actually this could be a good reason for separation. Themoviedb for example uses as year the first premiered date and then users can provide localized premiere date in addition so the addon could provide them.

**Comment by phate89** [comment]
updated the pr. I found a way to keep one field and avoid to fill the date with dummy dates.
ma premiered and m_bHasPremiered should be private because they can only be modified from members to keep them reliable but I can't because of cxx stuff.
@Montellese when you find some time can you please review it again?
@HitcherUK with latest changes the premiered label is filled only if there's a real full date. If there's not it's empty. The year part works as before.

**Comment by Razzeee** [comment]
So if I read this right the worst thing that can happen is an addon/scraper trying to set the year and it will be ignored by core when premiered is already set? Seems like a solution, might cause some inconveniences but shouldn't be a problem.

**Comment by phate89** [comment]
If an anddon/scraper/json set both only the premiered is considered. If an addon sets only the year the year is replaced in the current existing date (if exists) to avoid loosing it

**Comment by phate89** [comment]
rebased
@Montellese @Razzeee when you find some time can you look if this solution is ok?

File: xbmc/video/VideoDatabase.cpp [hunk 2]
```diff
@@ -2072,6 +2072,8 @@ std::string CVideoDatabase::GetValueString(const CVideoInfoTag &details, int min
   std::vector<std::string> conditions;
   for (int i = min + 1; i < max; ++i)
   {
+    if (offsets[i].type == NULL || offsets[i].offset == NULL)
+      continue;
```
Comment (Montellese) [inline xbmc/video/VideoDatabase.cpp]: Did you run into any issues here?

File: xbmc/video/VideoDatabase.cpp [hunk 3]
```diff
@@ -2236,6 +2238,10 @@ int CVideoDatabase::SetDetailsForMovie(const std::string& strFilenameAndPath, CV
       sql += PrepareSQL(", userrating = %i", details.m_iUserRating);
     else
       sql += ", userrating = NULL";
+    if (details.HasPremiered())
+      sql += PrepareSQL(", premiered = '%s'", details.GetPremiered().GetAsDBDate().c_str());
+    else
+      sql += PrepareSQL(", premiered = '%i'", details.GetYear());
```
Comment (Montellese) [inline xbmc/video/VideoDatabase.cpp]: Why not write a string in the form of `<year>-01-01` (or whatever the DB date format looks like)? That way you don't need to do any odd handling when reading the values again.

File: xbmc/video/VideoDatabase.h [hunk 4]
```diff
@@ -198,10 +200,10 @@ const struct SDbTableOffsets
   { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strPlot) },
   { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strPlotOutline) },
   { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strTagLine) },
-  { VIDEODB_TYPE_STRING, my_offsetof(CVideoInfoTag,m_strVotes) },
+  { NULL, NULL }, // unused
```
Comment (Montellese) [inline xbmc/video/VideoDatabase.h]: This is unrelated (same below). Please move it into a separate commit.

File: xbmc/video/VideoInfoTag.h [hunk 5]
```diff
@@ -157,7 +164,6 @@ class CVideoInfoTag : public IArchivable, public ISerializable, public ISortable
   CScraperUrl m_strPictureURL;
   std::string m_strTitle;
   std::string m_strSortTitle;
-  std::string m_strVotes;
```
Comment (Montellese) [inline xbmc/video/VideoInfoTag.h]: Unrelated?

**Comment by Montellese** [comment]
Is there any reason why you are not reusing the existing column that is currently used for the year? You could simply rename the enum value instead of marking it as unused.

File: xbmc/video/VideoInfoTag.cpp [hunk 6]
```diff
@@ -219,7 +217,10 @@ bool CVideoInfoTag::Save(TiXmlNode *node, const std::string &tag, bool savePathI
   XMLUtils::SetStringArray(movie, "tag", m_tags);
   XMLUtils::SetStringArray(movie, "credits", m_writingCredits);
   XMLUtils::SetStringArray(movie, "director", m_director);
-  XMLUtils::SetDate(movie, "premiered", m_premiered);
+  if (HasPremiered())
+    XMLUtils::SetDate(movie, "premiered", m_premiered);
+  else if (HasYear())
```
Comment (Montellese) [inline xbmc/video/VideoInfoTag.cpp]: Could we write both to keep backwards compatibility?

File: xbmc/video/VideoInfoTag.cpp [hunk 7]
```diff
@@ -549,7 +550,7 @@ void CVideoInfoTag::Serialize(CVariant& value) const
   value["playcount"] = m_playCount;
   value["lastplayed"] = m_lastPlayed.IsValid() ? m_lastPlayed.GetAsDBDateTime() : StringUtils::Empty;
   value["top250"] = m_iTop250;
-  value["year"] = m_iYear;
+  value["year"] = m_premiered.GetYear();
```
Comment (Montellese) [inline xbmc/video/VideoInfoTag.cpp]: What does `CDateTime::GetYear()` return if the `CDateTime` object is invalid? It must be `0` or it needs to be properly handled here.

**Comment by phate89** [comment]
The main reason is that there's no way to know if there is a premiere date or just the year. When you upgrade you will have all xxxx-01-01 dates and it's ugly. Also some scraper only have the year so if they set the year they will only get fake dates. With this solution the premiere date string will be empty for skins if we only have the year.
I added a new field to the db because of this solution since with the cXX stuff I can't control the data it save to the db..the alternative was to have 2 fields and store the 2 results separately...

File: xbmc/video/VideoDatabase.cpp (hunk 2 above)
Comment (phate89) [inline xbmc/video/VideoDatabase.cpp]: No issues. With this check I can set the offset to null so I can drop 2 variables that we constantly fill and carry around even if they're not used in CVideoInfoTag.  
That's unrelated I'll split it in a separate commit

File: xbmc/video/VideoInfoTag.cpp (hunk 7 above)
Comment (phate89) [inline xbmc/video/VideoInfoTag.cpp]: getyear do a check if the date is valid and if it's not it returns 0

**Comment by phate89** [comment]
fixed the xml saving for backward compatibility and separated the unrelated changes in a different commit.

**Comment by phil65** [comment]
Estuary should already support this btw, in case you want something to test with.

**Comment by Tolriq** [comment]
@phate89 API is changed version must be 7.14 and can you please add a comment to setXXdetails about the fact that premiered have priority over year when both are set.

**Comment by phate89** [comment]
@Tolriq 
I never understood when we bump only minor version..
The comment is ok in the description?

**Comment by Tolriq** [comment]
The 13 to 14 is the minor ;) the 7 is the major, the last one is revision.

minor is API change without breaks, major is breaks, revision is usually no API change but something was fixed. In some rare case there was API change with revision only updated but this should not happens.

For description you can add one to the premiered and year field to say they are linked (only in setXXXDetails), for those kind of remark I don't know if global description should be used as it could end up quite long, but it's @Montellese call for that.

**Comment by phate89** [comment]
@Tolriq 
thanks.
I added a description to the field.. If you or anyone have better words to express it suggestions are welcome :)

**Comment by Tolriq** [comment]
7.14**.0** :) 

About wording I'm not the correct one for that :p But at least put Uppercase to first Letters of sentence and after the dot.

**Comment by phate89** [comment]
if @Razzeee  and @Montellese aren't against this method I'll merge it
jenkins build this please

**Comment by Montellese** [comment]
No objections from me.

**Comment by phate89** [comment]
jenkins build this please

**Comment by fritsch** [comment]
@phate89 

The NULL changes won't work as intended:

> /home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:223:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
 };
 ^
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:223:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:223:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:223:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:265:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
 };
 ^
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:265:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:339:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
 };
 ^
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:339:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:339:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:339:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:379:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
 };
 ^
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:379:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:379:1: warning: converting to non-pointer type ‘int’ from NULL [-Wconversion-null]
/home/fritsch/Desktop/xbmc-fritsch/xbmc/xbmc/video/VideoDatabase.h:379:1: warning: converting to non-pointer type ‘size_t {aka long unsigned int}’ from NULL [-Wconversion-null]

**Comment by fritsch** [comment]
Basically keep it as it was and change they key to something _UNUSED and the value is not allowed to be NULL, too. This is directly given into a macro and casted to int later on. I am quite sure this deprecation does not work as intended. As you want the offset to be 0 (not NULL) change it to 0.

@Montellese @tamland you gave your +1 - what was the intention?

**Comment by phate89** [comment]
@fritsch 
I checked in debug and it works. NULL is changed to 0 by the compiler and 0 is not a valid VIDEODB_TYPE (they start from 1). So when I check if it's equal to null it compares to 0 and it doesn't parse the field..
When I find time I'll change it to something cleaner (maybe I'll add a VIDEODB_TYPE_NO_FIELD as 0)

**Comment by fritsch** [comment]
See above warnings :-) I did not write them myself :-)

What about: http://paste.ubuntu.com/17026050/

**Comment by phate89** [comment]
I'm not questioning that isn't a great implementation that create warnings.. but the end compiled result is working as it should ignoring that fields...
Anyway thanks for the code. It's what I had in mind.. I'll create a pr this evening

**Comment by fritsch** [comment]
@phate89 we are currently (see the fix 1000 warnings PR) to get kodi building without warnings to see the real warnings. In order to not have to do housekeeping _afterwards_ let's not make the codebase dirty while writing the first iteration of code :-)

Thanks much.

**Comment by phate89** [comment]
@fritsch I perfectly understand that..
only "The NULL changes won't work as intended" isn't true.. It works as intended but with terrible code :;)

**Comment by axbmcuser** [comment]
@phate89 
Great work! Looks like there is still a bug with some items randomly failing to scan the full premiered-date. Please see: http://forum.kodi.tv/showthread.php?tid=251922&pid=2357875#pid2357875
